src/Autofac.Extensions.DependencyInjection/AutofacServiceProvider.cs

```diff
@@ -24,27 +24,68 @@
 // OTHER DEALINGS IN THE SOFTWARE.
 
 using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
 using Microsoft.Extensions.DependencyInjection;
 
 namespace Autofac.Extensions.DependencyInjection
 {
     internal class AutofacServiceProvider : IServiceProvider, ISupportRequiredService
     {
         private readonly IComponentContext _componentContext;
+        private List<OrderedServiceDescriptor> _originalDescriptors;
+        private HashSet<Type> _serviceTypes;
+        private MethodInfo _castMethodInfo;
 
-        public AutofacServiceProvider(IComponentContext componentContext)
+        public AutofacServiceProvider(IComponentContext componentContext, List<ServiceDescriptor> originalDescriptors)
         {
             _componentContext = componentContext;
+            _originalDescriptors = originalDescriptors.Select((d, i) => new OrderedServiceDescriptor
+            {
+                Index = i,
+                ImplementationInstance = d.ImplementationInstance,
+                ImplementationType = d.ImplementationType
+            }).ToList();
+            _serviceTypes = new HashSet<Type>(originalDescriptors.Select(d => d.ServiceType));
+            _castMethodInfo = typeof(Enumerable).GetTypeInfo().GetDeclaredMethod("Cast");
         }
 
         public object GetService(Type serviceType)
         {
-            return _componentContext.ResolveOptional(serviceType);
+            return GetOrdered(serviceType, _componentContext.ResolveOptional(serviceType));
```

Comment:
@aL3891 This works in a Controller, because controllers are activated using `ActivatorUtilities`. It breaks down if you have a transitive dependencies.

For example, if `MyController` injects `MyService`, and `MyService` injects `IEnumerable<AnotherService>` you should see the issue.